### PR TITLE
feat(server) Start an HTTPS server (with self-signed certificate).

### PR DESCRIPTION
**Pull request by Hywan**
This patch introduces a `protocol` entry in the configuration
file (can be overriden by the `PROTOCOL` env variable). The
default value is `http`, which ensures the same behaviour
than usual. Changing it to `https` and an HTTPS server will start
instead of an HTTP one. A default self-signed certificate + key
are located in `config/server/`. A fresh self-signed certificate +
key will be generate on start up.

This feature is useful for Web Payment (`PaymentRequest` API, Apple
Pay, Google Pay, Samsung Pay etc.) where some API are only available
when the documents are served through HTTPS.

It has to be understood that this feature is for development only, and
probably restricted to _local_ development only. It is not aimed to be
used in production at all.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -5,7 +5,9 @@
  */
 var boot = require( 'boot' ),
 	http = require( 'http' ),
-	chalk = require( 'chalk' );
+	https = require( 'https' ),
```
Comment (blowery) [inline index.js]: might as well use `http2` here? https://nodejs.org/api/http2.html#http2_server_side_example

File: test/server/certificate.pem [hunk 2]
```diff
@@ -0,0 +1,16 @@
+-----BEGIN CERTIFICATE-----
```
Comment (blowery) [inline test/server/certificate.pem]: these should probably live under `config`, not `test`. ie `config/server/...`

Comment (Hywan) [reply test/server/certificate.pem]: True!

File: index.js (hunk 1 above)
Comment (Hywan) [reply index.js]: Do you know if `http2` is compatible with `http` or `https`?

Comment (Hywan) [reply index.js]: It seems to be compatible 😃.

Comment (Hywan) [reply index.js]: … in appearance only. I've errors when using the `http2` package. I would stick with `https` for now, and see if we need `http2` later, that's not my priority for now :smiley:.

**Comment by jsnajdr** [comment]
Could we also support an environment variable that tells Calypso server to run as HTTP?
```bash
CALYPSO_PROTOCOL=https npm run start
```
That's a bit better developer experience in dev mode: I don't have to edit the `config/development.json` file, which is version controlled, and be on constant vigilance to not accidentally commit the changes...

I'm also somewhat paranoid about publishing an actual self-signed certificate in a Git repo. After I add a security exception for this cert for the `calypso.localhost` domain, suddenly my browser trusts a certificate whose private key is public... I don't know if it's possible to exploit that, but it sounds a bit scary.

Could we make everyone generate their own self-signed cert? It could be automated by the `npm start` script. The `openssl` command needed to generate a cert is available on all dev platforms we support.

**Comment by Hywan** [comment]
@jsnajdr You can always use `PROTOCOL=https npm run start` to switch to HTTPS, just like `HOST=… PORT=… npm run start`.

I can try to generate a certificate with `npm start` :-).

**Comment by Hywan** [comment]
@jsnajdr Done :-). Hope it makes you less paranoid and brings you more comfort ;-).

File: index.js [hunk 3]
```diff
@@ -29,9 +30,45 @@ function sendBootStatus( status ) {
...
+	let certificate;
+
+	try {
+		execSync( 'openssl version', execOptions );
+		execSync( 'openssl req -x509 -newkey rsa:2048 -keyout ./config/server/key.tmp.pem -out ./config/server/certificate.pem -days 365 -nodes -subj "/C=US/ST=Foo/L=Bar/O=Baz/CN=calypso.localhost"', execOptions );
+		execSync( 'openssl rsa -in ./config/server/key.tmp.pem -out ./config/server/key.pem', execOptions );
+		execSync( 'rm ./config/server/key.tmp.pem', execOptions );
+
+		key = './config/server/key.pem';
+		certificate = './config/server/certificate.pem';
```
Comment (jsnajdr) [inline index.js]: Can we check if the local `key.pem` and `certificate.pem` already exist before generating them? Currently, a new certificate is generated on every `npm start` and I need to confirm a security exception in the browser every time.

Other than that, the HTTPS server works 👌 Thanks for making it happen!

Comment (Hywan) [reply index.js]: Sure. Will do that on Monday :-).

Comment (Hywan) [reply index.js]: Done!

File: index.js [hunk 4]
```diff
@@ -29,9 +30,44 @@ function sendBootStatus( status ) {
...
 
-server = http.createServer( app );
+// Start a development HTTPS server.
+if ( protocol === 'https' ) {
+	const { execSync } = require( 'child_process' );
+	const execOptions = { encoding: 'utf-8', windowsHide: true };
+	let key = './config/server/key.pem';
+	let certificate = './config/server/certificate.pem';
+
+	if ( ! fs.existsSync( './config/server/key.pem' ) || ! fs.existsSync( './config/server/certificate.pem' ) ) {
```
Comment (jsnajdr) [inline index.js]: nit: you can use the `key` and `certificate` variables here.

Comment (Hywan) [reply index.js]: Done.

**Comment by Hywan** [comment]
> I'd consider removing the default key and cert altogether and use only the generated one, but that's completely optional

In case OpenSSL fails, I think it's better for the user experience to keep the default key and certificate files. The security impact you rightly mentionned seems very low in such a situation.

Merging :-).

**Pull request merged**